### PR TITLE
MLv2 metadata calculation

### LINKED ISSUE #28719: [MLv2] Recursively infer `:lib/stage-metadata` for joins
In https://github.com/metabase/metabase/issues/28690 we started adding `:lib/stage-metadata` to each stage of the pMBQL query. However we're not yet doing that for the `:joins` in each stage. 

Since `:joins` also have `:stages` just like the query as a whole it should be easy enough to add this stuff.

This should be done automatically in the `metabase.lib.metadata.columns` functions or wherever it is that we're adding this stuff to the top-level stages when queries come in from `metabase.lib.query`

### LINKED ISSUE #28883: [MLv2] Functions for calculating metadata at any stage of a query
The basic idea is that given the `DatabaseMetadataProvider` implemented in #28881 (and `:lib/stage-metadata` if the first stage is a native query), we should be able to infer the metadata for all of the subsequent stages. We already have this Advanced Technology in `metabase.query-processor.middleware.annotate`, so it's really just a matter of porting it to Cljc.

Mega bonus points if we can write a QP Store `DatabaseMetadataProvider` and the `annotate` middleware can use the new MLv2 version of the code.

I started working on this in #28719, but went with a different approach for the metadata overhaul

### PR DESCRIPTION
**Pull request by camsaul**
Resolves #28883 
Resolves #28719

The humanization-related changes are split off into #29009, so you can ignore those here.

This PR introduces a new namespace, `metabase.lib.calculate`, that can calculate metadata about the columns returned by any stage of a query, by looking at the metadata for the previous stage, "visible" columns in the source table, results metadata for native queries or other source queries, joins, etc.

It's heavily inspired by `metabase.query-processor.middleware.annotate`, originally I was going to try to rewrite that namespace to use this new code, but that ended up being a little more work than I thought and this PR is big enough as it is. We can do that at some point in the future tho, it would be nice to have extra coverage of this code.

This also has sophisticated logic for calculating expression types and generating nice display names, similar to code in annotate but actually more sophisticated than the old stuff. 

For example if you write a kooky expression like 

```clj
[:+
 {}
 [:min {} [:field {} (meta/id :venues :id)]]
 [:*
  {}
  2
  [:avg {} [:field {} (meta/id :venues :price)]]
  3
  [:- {} [:max {} [:field {} (meta/id :venues :category-id)]] 4]]]
```

It will generate the display name

```
Min of ID + (2 × Average of Price × 3 × (Max of Category ID - 4))
```

neat!

It can also infer the type of that expression. `:type/Float` in this case, but if you were to remove that `:avg` aggregation it's smart enough to realize that `:+`, `:-`, and `:*` with all integers will return integers and calculate the result type of the entire expression as `:type/Integer`. N E A T

Note that it doesn't even need `:base-type` information attached to the `:field-clauses`, it can use the metadata provider to look this stuff up. 

What will we do with this cool new technology? The field calculation is a necessary prerequisite  for stuff like 

- https://github.com/metabase/metabase/issues/28703 
- https://github.com/metabase/metabase/issues/28702

but the display name and type calculation stuff can probably be put to good use for other purposes as well

### CODE REVIEW

**Review by bshepherdson — APPROVED** [review]

File: src/metabase/lib/metadata/calculate.cljc [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+           :field_ref aggregation-ref)))
+
+(mu/defn ^:private metadata-for-ref :- lib.metadata/ColumnMetadata
+  [query        :- ::lib.schema/query
+   stage-number :- :int
+   ref          :- ::lib.schema.ref/ref]
+  (case (first ref)
+    :field       (metadata-for-field-ref query stage-number ref)
+    :expression  (metadata-for-expression-ref query stage-number ref)
+    :aggregation (metadata-for-aggregation-ref query stage-number ref)))
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/lib/metadata/calculate.cljc]: good point. Fixed

File: src/metabase/lib/metadata/calculate/resolve.cljc [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+(mu/defn join :- ::lib.schema.join/join
+  "Resolve a join with a specific `join-alias`."
+  [query        :- ::lib.schema/query
+   stage-number :- :int
+   join-alias   :- ::lib.schema.common/non-blank-string]
+  (or (some (fn [join-map]
+              (when (= (:alias join-map) join-alias)
+                join-map))
+            (:joins (lib.util/query-stage query stage-number)))
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/lib/metadata/calculate/resolve.cljc:62]: Medley.core has `find-first` actually, I guess we should probably use that

[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/lib/metadata/calculate/resolve.cljc:62]: I'll update the code to use it

File: src/metabase/lib/metadata/calculate/type.cljc [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+(defmulti ^:private base-type*
+  {:arglists '([query stage-number expr])}
+  (fn [_query _stage-number expr]
+    (lib.dispatch/dispatch-value expr)))
+
+(defmethod base-type* :default
+  [_query _stage-number expr]
+  (or (some (fn [[expression-type-schema base-type]]
+              (when (mc/validate expression-type-schema expr)
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/lib/metadata/calculate/type.cljc]: Ended up removing this code

File: src/metabase/lib/metadata/calculate/type.cljc [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+  [_query _stage-number _clause]
+  :type/Text)
+
+(defmethod base-type* :rtrim
+  [_query _stage-number _clause]
+  :type/Text)
+
+(defmethod base-type* :length
+  [_query _stage-number _clause]
+  :type/Integer)
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/lib/metadata/calculate/type.cljc]: Good point. I ended up removing all of this stuff anyway, the new `mbql-clause/define-mbql-clause` stuff lets you specify the type stuff inline

File: test/metabase/lib/metadata/calculate/names_test.cljc [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+   [metabase.lib.test-metadata :as meta]))
+
+(def ^:private venues-query
+  {:lib/type     :mbql/query
+   :lib/metadata meta/metadata-provider
+   :type         :pipeline
+   :database     (meta/id)
+   :stages       [{:lib/type     :mbql.stage/mbql
+                   :lib/options  {:lib/uuid (str (random-uuid))}
+                   :source-table (meta/id :venues)}]})
```
[the remark this answers is not in the document]
Comment (camsaul) [reply test/metabase/lib/metadata/calculate/names_test.cljc:15]: good idea

File: src/metabase/lib/metadata/calculate/names.cljc [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+(mu/defn column-name :- ::lib.schema.common/non-blank-string
+  "Calculate a database-friendly name to use for an expression."
+  [query                                    :- ::lib.schema/query
+   stage-number                             :- :int
+   x]
+  (or
+   ;; if this is an MBQL clause with `:name` in the options map, then use that rather than calculating a name.
+   (when (and (vector? x)
+              (keyword? (first x))
+              (map? (second x)))
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/lib/metadata/calculate/names.cljc:63]: fixed

**Pull request merged**